### DISCUSSION TITLE
Reduced Integration in MOOSE

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#20750] [Q&A General] [answered]
Hello,

is there a way to find out how many quadrature points an element uses in MOOSE? For example ABAQUS has a documentation for each element where the nodes and the quadrature points are numbered respectively. Looking through the LibMesh documentation I only found a numbering for the nodes but not for the qps. Does that mean every element in MOOSE is fully integrated? Is there a way to manually switch to reduced integration?

I'm aware there is the `volumetric_locking_correction` feature but to my understanding that is not same thing. Or am I wrong here?

**Comment by recuero** [comment, marked as answer]
Elements are by default integrated via the exact quadrature rule. This can be changed from the quadrature executioner block.

The volumetric locking correction flag is recommended for mechanics problems and first order Lagrange elements.

**Reply by recuero** [reply]
If you set that flag, spatial numerical integration is automatically changed to remove volumetric locking and distinguish between deviatoric and volumetric parts. Which is a particular way to do selective integration.

**Reply by GiudGiud** [reply]
Hello

You can set the quadrature block in the Executioner block as mentioned above, see this page:
https://mooseframework.inl.gov/source/actions/SetupQuadratureAction.html

Guillaume

**Reply by jessecarterMOOSE** [reply]
Can I ask a follow up question? What is the "order" (first, second, etc...) nomenclature mean in the Quadrature block?

For instance, if I have second-order (n=2) variables, and I want to integrate them fully, is the Quadrature order "second" to match the order of the polynomial? Or does it refer to the number of quadrature points?

**Reply by GiudGiud** [reply]
so it's linked to the order of the polynomial (usually) basis the quadrature is obtained from. 

For a second order variable, you have to multiply it by a second order test function on top so you end up with needing a 4th order quadrature to integrate it perfectly (well I m taking a lot of shortcuts here, depends on the quadrature).

**Reply by jessecarterMOOSE** [reply]
Well let's assume Gauss quadrature for now. In Abaqus terminology, reduced integration means "elements use one fewer integration point in each direction", [see here](https://abaqus-docs.mit.edu/2017/English/SIMACAEGSARefMap/simagsa-c-ctmreduced.htm). In that link, the figure on the right is a 2D second-order "reduced" element with only 2 quadrature points in each dimension (4 total), as opposed to the typical "fully-integrated" element like that [seen here](https://abaqus-docs.mit.edu/2017/English/SIMACAEGSARefMap/simagsa-c-ctmfull.htm) with 3 quadrature points in each direction (9 total).

How would one implement the "reduced" element in MOOSE? I think the element type would be HEX8, variables would be second order Lagrange, and the quadrature order would be third? Would anything else change?

**Reply by GiudGiud** [reply]
Would you want to reduce the order for an element or just for a side?
Either way the parameters to do this are here
https://mooseframework.inl.gov/source/actions/SetupQuadratureAction.html

You dont need to downgrade the element type IMO, just the quadrature.

**Reply by jessecarterMOOSE** [reply]
Whoops, I meant QUAD8 element, not HEX8. Sorry. Trying to reduce the quadrature inside the element. Basically just like the examples I linked to.

**Reply by GiudGiud** [reply]
if it's for the whole element not just the side I think you ll need to define a block. 
@recuero for correctness-check on this

**Reply by hugary1995** [reply]
Setting quadrature order to element order minus one is one way to do what you described. But that's not ideal in the multiphysics setting as moose doesn't allow different quadrature for different variables. Ideally you only want to do reduced integration for the mechanics problem, (if that abaqus page is what you are after) volumetric_locking_correction is the right parameter to set.

**Reply by recuero** [reply]
As ~@hugary~  @hugary1995 said, in a multiphysics run, the quadrature rule you choose in the executioner will apply to all variables in the element (per mesh block). You may get away with underintegrating in some multiphysics problems potentially, but I'd assume this will need to be assessed case by case. The particular modes of deformation of an element typically have a larger importance in mechanics problems with plasticity, e.g, compared to other physics.

I think the original question goes back to the fact that, to ensure correctness and optimal computational effort in solid mechanics finite element problems, reduced integration is typically used for higher order elements (for first order, we'd need hourglass controls, which we don't have yet). If you want to use higher order elements, I'd be hesitant to use the volumetric locking correction (maybe Gary has studied this more and can comment). Maybe it's worth playing with reduced integration with your problem. Based on Guillaume's description of libmesh's quadrature option, selecting `THIRD` and the blocks you want your quadrature to apply to will do the trick.

**Reply by jessecarterMOOSE** [reply]
OK I think this essentially reproduces the Abaqus 2D second order plane strain reduced integration element. The VPP writes out just 4 lines for each timestep corresponding to the 4 quadrature points, rather than 9 when using full (fourth-order) integration quadrature. I'm just applying a simple displacement to the top face.

```
[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 1
  ny = 1
  second_order = true
  elem_type = QUAD8
[]

[Variables]
  [./disp_x]
    order = SECOND
    family = LAGRANGE
  [../]
  [./disp_y]
    order = SECOND
    family = LAGRANGE
  [../]
[]

[Modules/TensorMechanics/Master]
  [./plane_strain]
    strain = SMALL
    out_of_plane_direction = z
    planar_formulation = PLANE_STRAIN
    generate_output = 'stress_xx stress_xy stress_yy stress_zz strain_xx strain_xy strain_yy strain_zz'
  [../]
[]

[Functions]
  [./pullfunc]
    type = ParsedFunction
    value = 't/100.0'
  [../]
[]

[BCs]
  [./leftx]
    type = DirichletBC
    boundary = left
    variable = disp_x
    value = 0.0
  [../]
  [./bottomy]
    type = DirichletBC
    boundary = bottom
    variable = disp_y
    value = 0.0
  [../]
  [./toppull]
    type = FunctionDirichletBC
    boundary = top
    variable = disp_y
    function = pullfunc
  [../]
[]

[Materials]
  [./elastic_stress]
    type = ComputeLinearElasticStress
  [../]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    poissons_ratio = 0.3
    youngs_modulus = 1e6
  [../]
  [mat]
    type = GenericFunctionMaterial
    prop_names = 'prop1'
    prop_values = '1'
  []
[]

[Executioner]
  type = Transient

  solve_type = PJFNK
  line_search = none
  l_max_its = 100
  l_tol = 1e-10
  nl_max_its = 10
  nl_rel_tol = 1e-12
  start_time = 0.0
  dt = 1.0
  num_steps = 2

  [Quadrature]
    order = third
  []
[]

[VectorPostprocessors]
  [vpp]
    type = MaterialVectorPostprocessor
    material = 'mat'
    elem_ids = '0'
    execute_on = 'initial'
  []
[]

[Postprocessors]
  [avg_stressyy]
    type = ElementAverageValue
    variable = stress_yy
  []
[]

[Outputs]
  csv = true
  [./exodus]
    type = Exodus
  [../]
[]
```

**Reply by jessecarterMOOSE** [reply]
Thanks all for your insights. Meant to say that I'm looking at second-order displacements. And I understand your point about different physics, but in my case I'm looking at just mechanics, essentially trying to reproduce Abaqus. Sorry to take over this thread. I was just confused as to the meaning of "order" in relation to shape functions and quadrature.
